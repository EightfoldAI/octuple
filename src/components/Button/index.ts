export * from './Button.types';
export * from './BaseButton';
export * from './DefaultButton/DefaultButton';
export * from './NeutralButton/NeutralButton';
export * from './PrimaryButton/PrimaryButton';
export * from './SecondaryButton/SecondaryButton';
export * from './SplitButton/SplitButton';
export * from './TwoStateButton/TwoStateButton';
