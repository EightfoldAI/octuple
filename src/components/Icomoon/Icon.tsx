import React, { FC, useContext } from 'react';
import { mergeClasses } from '../../shared/utilities';
import { IcomoonProps, IcomoonSize } from './';
import IcomoonReact from 'icomoon-react';

import styles from './icon.module.scss';

const IconSetContext = React.createContext<object | null>(null);

export const IcomoonProvider = IconSetContext.Provider;

export const Icomoon: FC<IcomoonProps> = ({
    ariaHidden = false,
    classNames,
    color,
    iconName,
    id,
    role = 'presentation',
    size = IcomoonSize.Medium,
    'data-test-id': dataTestId,
}) => {
    const iconClassNames: string = mergeClasses([
        classNames,
        styles.iconWrapper,
    ]);

    return (
        <span
            data-test-id={dataTestId}
            aria-hidden={ariaHidden}
            className={iconClassNames}
            id={id}
            role={role}
        >
            <IconSetContext.Consumer>
                {(iconSet) => (
                    <IcomoonReact
                        iconSet={iconSet}
                        icon={iconName}
                        size={size}
                        color={color}
                    />
                )}
            </IconSetContext.Consumer>
        </span>
    );
};
