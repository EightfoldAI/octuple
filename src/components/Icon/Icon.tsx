'use client';

import React, { FC } from 'react';
import IcomoonReact from 'icomoon-react';

import { mergeClasses } from '../../shared/utilities';
import { Icon as MdiIcon } from '@mdi/react';
import { IconProps, IconSize } from './index';

import styles from './icon.module.scss';
import { useConfig } from '../ConfigProvider';

export const Icon: FC<IconProps> = ({
  ariaHidden = false,
  classNames,
  color,
  description,
  horizontal,
  id,
  path,
  role = 'presentation',
  rotate,
  size = IconSize.Medium,
  spin,
  style,
  title,
  vertical,
  'data-test-id': dataTestId,
  icomoonIconName,
}) => {
  const { icomoonIconSet } = useConfig();

  const iconClassNames: string = mergeClasses([classNames, styles.iconWrapper]);

  if (!icomoonIconName && !path) {
    return null;
  }

  const iconComponent = icomoonIconName ? (
    <IcomoonReact
      iconSet={icomoonIconSet}
      size={size}
      color={color}
      icon={icomoonIconName}
      role="presentation"
    />
  ) : (
    <MdiIcon
      color={color}
      description={description}
      horizontal={horizontal}
      path={path}
      rotate={rotate}
      size={size}
      title={title}
      vertical={vertical}
      spin={spin}
    />
  );

  return (
    <span
      data-test-id={dataTestId}
      aria-hidden={ariaHidden}
      className={iconClassNames}
      id={id}
      role={role}
      style={style ? style : null}
    >
      {iconComponent}
    </span>
  );
};

Icon.displayName = 'Icon';
