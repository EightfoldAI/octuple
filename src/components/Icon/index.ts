export * from './Icon';
export * from './Icon.types';
export * from './mdi';
