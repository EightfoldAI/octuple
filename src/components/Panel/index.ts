export * from './Panel';
export * from './PanelHeader';
export * from './Panel.types';
