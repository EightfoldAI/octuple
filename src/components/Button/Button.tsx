import React, { FC, Ref, useContext, useEffect, useRef, useState } from 'react';
import DisabledContext, { Disabled } from '../ConfigProvider/DisabledContext';
import { SizeContext, Size } from '../ConfigProvider';
import {
  ButtonIconAlign,
  ButtonProps,
  ButtonShape,
  ButtonSize,
  ButtonTextAlign,
  ButtonVariant,
  ButtonWidth,
  SplitButton,
} from '.';
import { Icon, IconSize } from '../Icon';
import { Badge } from '../Badge';
import { InnerNudge, NudgeAnimation, NudgeProps } from './Nudge';
import { Breakpoints, useMatchMedia } from '../../hooks/useMatchMedia';
import { mergeClasses } from '../../shared/utilities';
import { Loader, LoaderSize } from '../Loader';
import { useCanvasDirection } from '../../hooks/useCanvasDirection';
import { useMergedRefs } from '../../hooks/useMergedRefs';
import { useNudge } from './Nudge/Hooks/useNudge';

import styles from './button.module.scss';

export const Button: FC<ButtonProps> = React.forwardRef(
  (props: ButtonProps, ref: Ref<HTMLButtonElement>) => {
    const {
      alignIcon = ButtonIconAlign.Left,
      alignText = ButtonTextAlign.Center,
      allowDisabledFocus = false,
      ariaLabel,
      buttonWidth = ButtonWidth.fitContent,
      checked = false,
      classNames,
      configContextProps = {
        noDisabledContext: false,
        noSizeContext: false,
      },
      counter,
      disabled = false,
      disruptive = false,
      dropShadow = false,
      floatingButtonProps,
      nudgeProps: defaultNudgeProps,
      htmlType,
      iconProps,
      id,
      loading = false,
      onClick,
      onContextMenu,
      prefixIconProps,
      shape = ButtonShape.Pill,
      size = ButtonSize.Medium,
      split,
      splitButtonChecked = false,
      splitButtonProps,
      style,
      text,
      toggle,
      transparent = false,
      type, // TODO: Remove in Octuple v3.0.0 and use `variant` only.
      variant = ButtonVariant.Default,
      ...rest
    } = props;
    const largeScreenActive: boolean = useMatchMedia(Breakpoints.Large);
    const mediumScreenActive: boolean = useMatchMedia(Breakpoints.Medium);
    const smallScreenActive: boolean = useMatchMedia(Breakpoints.Small);
    const xSmallScreenActive: boolean = useMatchMedia(Breakpoints.XSmall);

    const internalRef: React.MutableRefObject<HTMLButtonElement> =
      useRef<HTMLButtonElement>(null);

    const mergedRef = useMergedRefs(internalRef, ref);

    const htmlDir: string = useCanvasDirection();

    const contextuallyDisabled: Disabled = useContext(DisabledContext);
    const mergedDisabled: boolean = configContextProps.noDisabledContext
      ? disabled
      : contextuallyDisabled || disabled;

    const contextuallySized: Size = useContext(SizeContext);
    const mergedSize = configContextProps.noSizeContext
      ? size
      : contextuallySized || size;

    const counterExists: boolean = !!counter;
    const iconExists: boolean = !!iconProps;
    const prefixIconExists: boolean = !!prefixIconProps;
    const textExists: boolean = !!text;

    const [nudgeProps, setNudgeProps] = useState<NudgeProps>(defaultNudgeProps);
    const innerNudgeRef: React.MutableRefObject<HTMLSpanElement> =
      useRef<HTMLSpanElement>(null);

    useEffect(() => {
      setNudgeProps(
        props.nudgeProps
          ? props.nudgeProps
          : {
              animation: NudgeAnimation.Background,
              delay: 2000,
              iterations: 1,
              enabled: false,
            }
      );
    }, [nudgeProps?.enabled]);

    useNudge(disruptive, nudgeProps, [internalRef, innerNudgeRef], styles);

    // TODO: Remove in Octuple v3.0.0 and use `variant` only.
    // For now, if `type` has a value use it, else use `variant`.
    const mergedVariant: ButtonVariant = !!type ? type : variant;

    const buttonSharedClassNames: string = mergeClasses([
      classNames,
      styles.button,
      { [styles.buttonDefault]: mergedVariant === ButtonVariant.Default },
      { [styles.buttonNeutral]: mergedVariant === ButtonVariant.Neutral },
      { [styles.buttonPrimary]: mergedVariant === ButtonVariant.Primary },
      { [styles.buttonSecondary]: mergedVariant === ButtonVariant.Secondary },
      { [styles.buttonSystemUi]: mergedVariant === ButtonVariant.SystemUI },
      {
        [styles.buttonDisruptive]:
          disruptive && mergedVariant === ButtonVariant.Default,
      },
      {
        [styles.buttonPrimaryDisruptive]:
          disruptive && mergedVariant === ButtonVariant.Primary,
      },
      {
        [styles.buttonSecondaryDisruptive]:
          disruptive && mergedVariant === ButtonVariant.Secondary,
      },
      {
        [styles.transparent]:
          transparent && mergedVariant === ButtonVariant.SystemUI,
      },
      {
        [styles.buttonSmall]:
          mergedSize === ButtonSize.Flex && largeScreenActive,
      },
      {
        [styles.buttonMedium]:
          mergedSize === ButtonSize.Flex && mediumScreenActive,
      },
      {
        [styles.buttonMedium]:
          mergedSize === ButtonSize.Flex && smallScreenActive,
      },
      {
        [styles.buttonLarge]:
          mergedSize === ButtonSize.Flex && xSmallScreenActive,
      },
      { [styles.buttonLarge]: mergedSize === ButtonSize.Large },
      { [styles.buttonMedium]: mergedSize === ButtonSize.Medium },
      { [styles.buttonSmall]: mergedSize === ButtonSize.Small },
      { [styles.pillShape]: shape === ButtonShape.Pill },
      {
        [styles.roundShape]:
          shape === ButtonShape.Round && !split && !textExists,
      },
      { [styles.dropShadow]: dropShadow },
      { [styles.disabled]: allowDisabledFocus || mergedDisabled },
      { [styles.floating]: floatingButtonProps?.enabled },
      { [styles.buttonRtl]: htmlDir === 'rtl' },
      {
        [styles.buttonConic]:
          !disruptive &&
          nudgeProps?.enabled &&
          nudgeProps?.animation === NudgeAnimation.Conic,
      },
    ]);

    const buttonClassNames: string = mergeClasses([
      buttonSharedClassNames,
      { [styles.buttonStretch]: buttonWidth === ButtonWidth.fill },
      { [styles.splitLeft]: split },
      { [styles.left]: alignText === ButtonTextAlign.Left },
      { [styles.right]: alignText === ButtonTextAlign.Right },
      {
        [styles.iconLeft]: iconExists && alignIcon === ButtonIconAlign.Left,
      },
      {
        [styles.iconRight]: iconExists && alignIcon === ButtonIconAlign.Right,
      },
      { [styles.loading]: loading },
    ]);

    const buttonTextClassNames: string = mergeClasses([
      {
        [styles.buttonTextSmall]:
          mergedSize === ButtonSize.Flex && largeScreenActive,
      },
      {
        [styles.buttonTextMedium]:
          mergedSize === ButtonSize.Flex && mediumScreenActive,
      },
      {
        [styles.buttonTextMedium]:
          mergedSize === ButtonSize.Flex && smallScreenActive,
      },
      {
        [styles.buttonTextLarge]:
          mergedSize === ButtonSize.Flex && xSmallScreenActive,
      },
      { [styles.buttonTextLarge]: mergedSize === ButtonSize.Large },
      { [styles.buttonTextMedium]: mergedSize === ButtonSize.Medium },
      { [styles.buttonTextSmall]: mergedSize === ButtonSize.Small },
    ]);

    const badgeClassNames: string = mergeClasses([
      styles.counter,
      buttonTextClassNames,
    ]);

    const getButtonIconSize = (): IconSize => {
      let iconSize: IconSize;
      if (mergedSize === ButtonSize.Flex && largeScreenActive) {
        iconSize = IconSize.Small;
      } else if (
        mergedSize === ButtonSize.Flex &&
        (mediumScreenActive || smallScreenActive)
      ) {
        iconSize = IconSize.Medium;
      } else if (mergedSize === ButtonSize.Flex && xSmallScreenActive) {
        iconSize = IconSize.Large;
      } else if (mergedSize === ButtonSize.Large) {
        iconSize = IconSize.Large;
      } else if (mergedSize === ButtonSize.Medium) {
        iconSize = IconSize.Medium;
      } else if (mergedSize === ButtonSize.Small) {
        iconSize = IconSize.Small;
      }
      return iconSize;
    };

    const getLoaderSize = (): LoaderSize => {
      let loaderSize: LoaderSize;
      if (size === ButtonSize.Flex && largeScreenActive) {
        loaderSize = LoaderSize.Small;
      } else if (
        size === ButtonSize.Flex &&
        (mediumScreenActive || smallScreenActive)
      ) {
        loaderSize = LoaderSize.Medium;
      } else if (size === ButtonSize.Flex && xSmallScreenActive) {
        loaderSize = LoaderSize.Large;
      } else if (size === ButtonSize.Large) {
        loaderSize = LoaderSize.Large;
      } else if (size === ButtonSize.Medium) {
        loaderSize = LoaderSize.Medium;
      } else if (size === ButtonSize.Small) {
        loaderSize = LoaderSize.Small;
      }
      return loaderSize;
    };

    const getButtonLoader = (): JSX.Element =>
      loading && (
        <Loader
          classNames={styles.loader}
          dotClassNames={styles.loaderDot}
          size={getLoaderSize()}
        />
      );

    const getButtonIcon = (): JSX.Element => (
      <Icon
        {...iconProps}
        classNames={mergeClasses([styles.icon, iconProps.classNames])}
        size={getButtonIconSize()}
      />
    );

    const getPrefixIcon = (): JSX.Element => (
      <Icon
        {...prefixIconProps}
        classNames={mergeClasses([
          styles.icon,
          styles.prefixIcon,
          prefixIconProps.classNames,
        ])}
        size={getButtonIconSize()}
      />
    );

    const getButtonContent = (
      buttonTextClassNames: string,
      text: string
    ): JSX.Element => (
      <span className={buttonTextClassNames}>
        {text ? text : 'Button'}
        {counterExists && <Badge classNames={badgeClassNames}>{counter}</Badge>}
      </span>
    );

    return (
      <>
        <button
          {...rest}
          ref={mergedRef}
          aria-checked={toggle ? !!checked : undefined}
          aria-disabled={mergedDisabled || loading}
          aria-label={ariaLabel}
          aria-pressed={toggle ? !!checked : undefined}
          defaultChecked={checked}
          disabled={(!allowDisabledFocus && mergedDisabled) || loading}
          className={buttonClassNames}
          id={id}
          onClick={!allowDisabledFocus ? onClick : null}
          style={style}
          type={htmlType}
        >
          <InnerNudge
            classNames={mergeClasses([
              styles.innerNudge,
              {
                [styles.conic]: nudgeProps?.animation === NudgeAnimation.Conic,
              },
            ])}
            disruptive={disruptive}
            id={id ? `${id}-nudge` : 'button-nudge'}
            nudgeProps={nudgeProps}
            ref={innerNudgeRef}
            style={style}
          />
          {iconExists && prefixIconExists && !textExists && (
            <span>
              {getButtonIcon()}
              {getPrefixIcon()}
            </span>
          )}
          {iconExists &&
            !textExists &&
            !prefixIconExists &&
            !counterExists &&
            getButtonIcon()}
          {counterExists && !textExists && !loading && !iconExists && (
            <Badge classNames={badgeClassNames}>{counter}</Badge>
          )}
          {iconExists && counterExists && !textExists && !loading && (
            <span>
              {getButtonIcon()}
              <Badge classNames={badgeClassNames}>{counter}</Badge>
              {prefixIconExists && getPrefixIcon()}
            </span>
          )}
          {iconExists && textExists && (
            <span>
              {getButtonIcon()}
              {getButtonContent(buttonTextClassNames, text)}
              {prefixIconExists && getPrefixIcon()}
            </span>
          )}
          {!iconExists && getButtonContent(buttonTextClassNames, text)}
          {getButtonLoader()}
        </button>
        {split && (
          <SplitButton
            {...splitButtonProps}
            classNames={
              buttonSharedClassNames + ' ' + splitButtonProps?.classNames
            }
            checked={splitButtonChecked}
            disruptive={disruptive}
            dropShadow={dropShadow}
            nudgeProps={nudgeProps}
            onClick={
              !splitButtonProps?.allowDisabledFocus ? onContextMenu : null
            }
            shape={shape}
            size={mergedSize}
            split={split}
            variant={mergedVariant}
          />
        )}
      </>
    );
  }
);
