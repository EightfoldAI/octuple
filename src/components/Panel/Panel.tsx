'use client';

import React, {
  useContext,
  useEffect,
  useImperativeHandle,
  useRef,
  useState,
} from 'react';
import GradientContext, { Gradient } from '../ConfigProvider/GradientContext';
import { OcThemeName } from '../ConfigProvider';
import ThemeContext, {
  ThemeContextProvider,
} from '../ConfigProvider/ThemeContext';
import { PanelLocale, PanelProps, PanelRef, PanelSize } from './';
import { Button, ButtonShape, ButtonVariant } from '../Button';
import { IconName } from '../Icon';
import { Portal } from '../Portal';
import { FocusTrap } from '../../shared/FocusTrap';
import { NoFormStyle } from '../Form/Context';
import { useCanvasDirection } from '../../hooks/useCanvasDirection';
import { useScrollLock } from '../../hooks/useScrollLock';
import { mergeClasses, stopPropagation } from '../../shared/utilities';
import LocaleReceiver, {
  useLocaleReceiver,
} from '../LocaleProvider/LocaleReceiver';
import enUS from './Locale/en_US';

import styles from './panel.module.scss';
import themedComponentStyles from './panel.theme.module.scss';

const PanelContext = React.createContext<PanelRef | null>(null);

const PANEL_WIDTHS: Record<PanelSize, number> = Object.freeze({
  small: 480,
  medium: 640,
  large: 860,
});

export const Panel = React.forwardRef<PanelRef, PanelProps>(
  (props: PanelProps, ref: React.ForwardedRef<PanelRef>) => {
    const {
      actionButtonOneProps,
      actionButtonTwoProps,
      actionButtonThreeProps,
      autoFocus = true,
      bodyClassNames,
      bodyPadding = true,
      children,
      closable = true,
      closeButtonAriaLabelText: defaultCloseButtonAriaLabelText,
      closeButtonProps,
      closeIcon = IconName.mdiClose,
      configContextProps = {
        noGradientContext: false,
        noThemeContext: false,
      },
      firstFocusableSelector,
      footer,
      footerClassNames,
      gradient = false,
      headerButtonProps,
      headerClassNames,
      headerIcon = IconName.mdiArrowLeftThick,
      headerPadding = true,
      height,
      lastFocusableSelector,
      locale = enUS,
      maskClosable = true,
      onClose = () => {},
      onVisibleChange,
      overlay = true,
      zIndex,
      panelClassNames,
      panelStyle,
      panelWrapperClassNames,
      parent = typeof document !== 'undefined' ? document.body : null,
      placement = 'right',
      push = true,
      renderContentAlways = true,
      size = PanelSize.medium,
      skipFocusableSelectorsFromIndex,
      title,
      visible = false,
      width,
      panelHeader,
      scrollLock = true,
      theme,
      themeContainerId,
      focusTrap = true,
      ...rest
    } = props;

    const htmlDir: string = useCanvasDirection();

    const panelRef = useRef<HTMLDivElement>(null);
    const containerRef = useRef<HTMLDivElement>(null);
    const parentPanel = useContext<PanelRef>(PanelContext);
    const [internalPush, setPush] = useState<boolean>(false);

    useScrollLock(parent, !scrollLock ? false : visible);

    const contextualGradient: Gradient = useContext(GradientContext);
    const mergedGradient: boolean = configContextProps.noGradientContext
      ? gradient
      : contextualGradient || gradient;

    const contextualTheme: OcThemeName = useContext(ThemeContext);
    const mergedTheme: OcThemeName = configContextProps.noThemeContext
      ? theme
      : contextualTheme || theme;

    // ============================ Strings ===========================
    const [panelLocale] = useLocaleReceiver('Panel');
    let mergedLocale: PanelLocale;

    if (props.locale) {
      mergedLocale = props.locale;
    } else {
      mergedLocale = panelLocale || props.locale;
    }

    const [closeButtonAriaLabelText, setCloseButtonAriaLabelText] =
      useState<string>(defaultCloseButtonAriaLabelText);

    // Locs: if the prop isn't provided use the loc defaults.
    // If the mergedLocale is changed, update.
    useEffect(() => {
      setCloseButtonAriaLabelText(
        props.closeButtonAriaLabelText
          ? props.closeButtonAriaLabelText
          : mergedLocale.lang!.closeButtonAriaLabelText
      );
    }, [mergedLocale]);

    const panelBackdropClasses: string = mergeClasses([
      styles.panelBackdrop,
      panelWrapperClassNames,
      { [styles.visible]: visible },
      { [styles.modeless]: overlay === false },
      { [styles.modelessMask]: overlay === false && maskClosable },
      { [styles.panelBackdropRtl]: htmlDir === 'rtl' },
    ]);

    const panelClasses: string = mergeClasses([
      styles.panel,
      { [themedComponentStyles.theme]: mergedTheme },
      { [styles.noBodyPadding]: bodyPadding === false },
      { [styles.noHeaderPadding]: headerPadding === false },
      panelClassNames,
      { [styles.right]: placement === 'right' },
      { [styles.left]: placement === 'left' },
      { [styles.bottom]: placement === 'bottom' },
      { [styles.top]: placement === 'top' },
      { [styles.large]: size === PanelSize.large },
      { [styles.medium]: size === PanelSize.medium },
      { [styles.small]: size === PanelSize.small },
    ]);

    const bodyClasses: string = mergeClasses([styles.body, bodyClassNames]);

    const footerClasses: string = mergeClasses([
      styles.footer,
      footerClassNames,
    ]);

    const headerClasses: string = mergeClasses([
      styles.header,
      headerClassNames,
    ]);

    useEffect(() => {
      if (parentPanel) {
        if (visible) {
          parentPanel.push();
        } else {
          parentPanel.pull();
        }
      }
      if (autoFocus) {
        setTimeout(() => {
          containerRef.current?.focus();
        }, 200);
      }
      onVisibleChange?.(visible);
    }, [visible]);

    const getDefaultHeader = (): JSX.Element => (
      <div className={headerClasses}>
        <div>
          {headerButtonProps && (
            <Button
              classNames={styles.headerButton}
              configContextProps={configContextProps}
              gradient={mergedGradient}
              iconProps={{
                path: headerIcon,
              }}
              style={{
                transform: htmlDir === 'rtl' ? 'rotate(180deg)' : 'none',
              }}
              shape={ButtonShape.Round}
              theme={mergedTheme}
              themeContainerId={themeContainerId}
              variant={ButtonVariant.Neutral}
              {...headerButtonProps}
            />
          )}
          {title}
        </div>
        <span className={styles.headerButtons}>
          {actionButtonThreeProps && (
            <Button
              configContextProps={configContextProps}
              gradient={mergedGradient}
              shape={ButtonShape.Round}
              theme={mergedTheme}
              themeContainerId={themeContainerId}
              variant={ButtonVariant.Neutral}
              {...actionButtonThreeProps}
            />
          )}
          {actionButtonTwoProps && (
            <Button
              configContextProps={configContextProps}
              gradient={mergedGradient}
              shape={ButtonShape.Round}
              theme={mergedTheme}
              themeContainerId={themeContainerId}
              variant={ButtonVariant.Neutral}
              {...actionButtonTwoProps}
            />
          )}
          {actionButtonOneProps && (
            <Button
              configContextProps={configContextProps}
              gradient={mergedGradient}
              shape={ButtonShape.Round}
              theme={mergedTheme}
              themeContainerId={themeContainerId}
              variant={ButtonVariant.Neutral}
              {...actionButtonOneProps}
            />
          )}
          {closable && (
            <Button
              configContextProps={configContextProps}
              gradient={mergedGradient}
              iconProps={{ path: closeIcon }}
              ariaLabel={closeButtonAriaLabelText}
              onClick={onClose}
              shape={ButtonShape.Round}
              theme={mergedTheme}
              themeContainerId={themeContainerId}
              variant={ButtonVariant.Neutral}
              {...closeButtonProps}
            />
          )}
        </span>
      </div>
    );

    const getHeader = (): JSX.Element => {
      if (!!panelHeader) return panelHeader;
      return getDefaultHeader();
    };

    const getBody = (): JSX.Element => (
      <div className={bodyClasses}>{children}</div>
    );

    const getFooter = (): JSX.Element => (
      <div className={footerClasses}>{footer}</div>
    );

    const getPushTransform = (): React.CSSProperties => {
      const panelWidth: number = width || PANEL_WIDTHS[size];
      let distance: number = visible ? 0 : panelWidth;

      if (internalPush) {
        distance -= panelWidth / 3;
      }

      if (['left', 'right'].includes(placement)) {
        if (htmlDir === 'rtl') {
          return {
            transform: `translateX(${
              placement === 'right' ? -distance : distance
            }px)`,
          };
        } else {
          return {
            transform: `translateX(${
              placement === 'left' ? -distance : distance
            }px)`,
          };
        }
      } else if (['top', 'bottom'].includes(placement)) {
        return {
          transform: `translateY(${
            placement === 'top' ? -distance : distance
          }px)`,
        };
      }
      return {};
    };

    const getPanelStyle = (): React.CSSProperties => ({
      zIndex,
      ...getPushTransform(),
      ...panelStyle,
      height,
      width,
    });

    const operations = React.useMemo(
      () => ({
        push() {
          push && setPush(true);
        },
        pull() {
          push && setPush(false);
        },
      }),
      [push]
    );

    useImperativeHandle(ref, () => operations);

    const getPanel = (): JSX.Element => {
      return (
        <LocaleReceiver componentName={'Panel'} defaultLocale={enUS}>
          {(_contextLocale: PanelLocale) => {
            return (
              <PanelContext.Provider value={operations}>
                <NoFormStyle status override>
                  <FocusTrap
                    firstFocusableSelector={firstFocusableSelector}
                    lastFocusableSelector={lastFocusableSelector}
                    skipFocusableSelectorsFromIndex={
                      skipFocusableSelectorsFromIndex
                    }
                    trap={visible && focusTrap}
                    {...rest}
                    ref={containerRef}
                    classNames={panelBackdropClasses}
                    onClick={(e: React.MouseEvent<HTMLDivElement>) => {
                      maskClosable && onClose(e);
                    }}
                    aria-hidden={!visible}
                  >
                    <ThemeContextProvider
                      componentClassName={themedComponentStyles.theme}
                      containerId={themeContainerId}
                      theme={mergedTheme}
                    >
                      <div
                        ref={panelRef}
                        className={panelClasses}
                        onClick={stopPropagation}
                        style={getPanelStyle()}
                      >
                        {renderContentAlways && (
                          <>
                            {getHeader()}
                            {getBody()}
                            {!!footer && getFooter()}
                          </>
                        )}
                      </div>
                    </ThemeContextProvider>
                  </FocusTrap>
                </NoFormStyle>
              </PanelContext.Provider>
            );
          }}
        </LocaleReceiver>
      );
    };

    return <Portal getContainer={() => parent}>{getPanel()}</Portal>;
  }
);
