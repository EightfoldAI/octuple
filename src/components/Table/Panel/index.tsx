import React from 'react';
import { TitleProps } from './Panel.types';

function Panel({ classNames, children }: TitleProps) {
    return <div className={classNames}>{children}</div>;
}

export default Panel;
