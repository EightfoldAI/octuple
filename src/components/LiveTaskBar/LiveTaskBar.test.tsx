import React from 'react';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import MatchMediaMock from 'jest-matchmedia-mock';
import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import { LiveTaskBar } from './';
import { Button, ButtonVariant } from '../Button';
import { IconName } from '../Icon';
import { PanelSize } from '../Panel';
import { Placement } from '@floating-ui/react';

Enzyme.configure({ adapter: new Adapter() });

let matchMedia: any;

describe('LiveTaskBar', () => {
  beforeAll(() => {
    matchMedia = new MatchMediaMock();
  });

  afterEach(() => {
    matchMedia.clear();
  });

  test('Renders with various configurations', () => {
    // Test basic rendering
    const { container, rerender } = render(
      <LiveTaskBar tasks={[]} data-testid="test-live-task-bar" />
    );
    expect(
      container.querySelector('[data-testid="test-live-task-bar"]')
    ).toBeTruthy();

    // Test with tasks
    rerender(
      <LiveTaskBar
        tasks={[
          <Button
            key="task1"
            text="Task 1"
            variant={ButtonVariant.Secondary}
          />,
          <Button
            key="task2"
            text="Task 2"
            variant={ButtonVariant.Secondary}
          />,
          <Button
            key="task3"
            text="Task 3"
            variant={ButtonVariant.Secondary}
          />,
        ]}
        data-testid="test-live-task-bar"
      />
    );
    expect(container.querySelectorAll('.task-items button').length).toBe(3);

    // Test with custom header
    const CustomHeader = () => (
      <div className="custom-header">Custom Header</div>
    );

    rerender(
      <LiveTaskBar
        header={<CustomHeader />}
        tasks={[]}
        data-testid="test-live-task-bar"
      />
    );
    expect(container.querySelector('.custom-header')).toBeTruthy();

    // Test with aiAgent theme
    rerender(
      <LiveTaskBar
        theme="aiAgent"
        tasks={[
          <Button
            key="task1"
            text="Task 1"
            variant={ButtonVariant.Secondary}
          />,
        ]}
        data-testid="test-live-task-bar"
      />
    );
    expect(container.querySelector('.ai-agent')).toBeTruthy();

    // Test with custom styles and classNames
    rerender(
      <LiveTaskBar
        tasks={[]}
        style={{ color: '#000000' }}
        classNames="custom-class"
        data-testid="test-live-task-bar"
      />
    );

    const taskBar = container.querySelector('.live-task-bar');
    if (taskBar instanceof HTMLElement) {
      expect(taskBar.style.color).toBe('rgb(0, 0, 0)');
    }
    expect(container.querySelector('.custom-class')).toBeTruthy();

    // Create one comprehensive snapshot
    expect(container).toMatchSnapshot();
  });

  test('Panel functionality works correctly', async () => {
    const panelProps = {
      title: 'Test Panel',
      size: PanelSize.medium,
      headerIcon: IconName.mdiInformation,
      children: <div>Panel content</div>,
    };

    const panelButtonProps = {
      text: 'Open Panel',
      iconProps: { path: IconName.mdiLightbulb },
      'data-testid': 'panel-button',
    };

    render(
      <LiveTaskBar
        tasks={[]}
        panel={panelProps}
        panelButton={panelButtonProps}
        data-testid="test-live-task-bar"
      />
    );

    // Panel should be initially hidden - checking in the document body since Panel uses Portal
    expect(document.querySelector('.panel-backdrop.visible')).toBeFalsy();

    // Click the button to show the panel
    const button = screen.getByTestId('panel-button');
    fireEvent.click(button);

    // Wait for the panel to be visible in the DOM
    await waitFor(() => {
      const visiblePanel = document.querySelector('.panel-backdrop.visible');
      expect(visiblePanel).toBeTruthy();
    });

    // Click again to hide
    fireEvent.click(button);

    // Wait for the panel to be hidden
    await waitFor(() => {
      const hiddenPanelAgain = document.querySelector(
        '.panel-backdrop.visible'
      );
      expect(hiddenPanelAgain).toBeFalsy();
    });

    // Test panel close button functionality
    fireEvent.click(button); // Open panel again

    // Wait for the panel to be visible again
    await waitFor(() => {
      expect(document.querySelector('.panel-backdrop.visible')).toBeTruthy();
    });

    // Find and click the close button within the panel
    const closeButton = document.querySelector(
      '.panel button[aria-label="Close"]'
    );
    if (closeButton) {
      fireEvent.click(closeButton as HTMLElement);
    }

    // Wait for the panel to be hidden after clicking close
    await waitFor(() => {
      const hiddenPanelAfterClose = document.querySelector(
        '.panel-backdrop.visible'
      );
      expect(hiddenPanelAfterClose).toBeFalsy();
    });
  });

  test('Dropdown functionality works correctly', () => {
    const dropdownProps = {
      placement: 'bottom-end' as Placement,
      trigger: 'click' as const,
      overlay: (
        <div className="dropdown-content">
          <button>Option 1</button>
          <button>Option 2</button>
        </div>
      ),
    };

    const dropdownButtonProps = {
      iconProps: { path: IconName.mdiDotsHorizontal },
      'data-testid': 'custom-dropdown-button',
      style: { backgroundColor: 'red' },
    };

    const { container } = render(
      <LiveTaskBar
        tasks={[]}
        dropdown={dropdownProps}
        dropdownButton={dropdownButtonProps}
        data-testid="test-live-task-bar"
      />
    );

    // Verify the custom dropdown button is rendered
    const dropdownButton = screen.getByTestId('custom-dropdown-button');
    expect(dropdownButton).toBeTruthy();

    // Verify the custom style is applied
    if (dropdownButton instanceof HTMLElement) {
      expect(dropdownButton.style.backgroundColor).toBe('red');
    }
  });
});
