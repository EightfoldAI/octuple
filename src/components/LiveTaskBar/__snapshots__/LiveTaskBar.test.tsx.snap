// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiveTaskBar Renders with various configurations 1`] = `
<div>
  <div
    aria-label="Live task status"
    class="live-task-bar-wrapper"
    data-testid="test-live-task-bar"
    role="region"
  >
    <div
      class="live-task-bar horizontal custom-class"
      style="color: rgb(0, 0, 0);"
    >
      <div
        class="left-section"
      />
      <div
        aria-label="Action items"
        class="action-items"
        role="group"
      />
    </div>
  </div>
</div>
`;
